### DISCUSSION TITLE
/the-basics/routing.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#269] [Announcements]
# /the-basics/routing.html

[[toc]] 简介 Goravel 路由模块可以使用 facades.Route() 进行操作。 HTTP 驱动 Goravel 默认使用 gin (https://github.com/gin-gonic/gin) 作为 HTTP 驱动，如果想使用其他驱动，可以到 config/http.go 中进行配置，目前官方默认支持 gin (https:/...

https://www.goravel.dev/zh/the-basics/routing.html

**Comment by luscio** [comment]
路由参数的类型，能像laravel那样正则约束吗？

**Reply by hwbrzzl** [reply]
目前还不行，目前可以在中间件中进行校验。

**Comment by islamsamy214** [comment]
facades.Route().RunTLS() doesn't work, because you have to add a certificate, but works with facades.Route().RunTLSWithCert("127.0.0.1:3000", "ca.pem", "ca.key"), So I think its better to remove it!?

**Reply by hwbrzzl** [reply]
Do you add the config? `config.GetString("http.tls.ssl.cert")` and `config.GetString("http.tls.ssl.key")`

**Reply by islamsamy214** [reply]
NOpe!, But I think it is worth mentioning in the docs, isn't it

**Reply by hwbrzzl** [reply]
We have it: https://www.goravel.dev/the-basics/routing.html#start-https-server

<img width="772" alt="image" src="https://github.com/goravel/goravel/assets/24771476/3555fbb6-8523-4a90-8b36-251e0c15f4f1">

**Reply by islamsamy214** [reply]
Yeah my bad!, thanks

**Comment by dimaspurwadi** [comment]
hi, why when i try to use

```
facades.Route().Group(func(router route.Router) {
  router.Get("group/{id}", func(ctx http.Context) http.Response {
    return ctx.Response().Success().String(ctx.Request().Query("id", "1"))
  })
})
```

i have error in part
```
route.Router
```
undefine route, which lib i need to import? i'm using v.1.13

Thanks

**Reply by hwbrzzl** [reply]
Yes, you should import `"github.com/goravel/framework/contracts/route"`

**Reply by hwbrzzl** [reply]
IED should import it automatically.

**Reply by dimaspurwadi** [reply]
ah thanks, i dont know why my IDE not import automatically.
btw can we include prefix in group? previously in my php laravel route, we can give prefix, for example:

$router->group(['prefix' => 'users'], function () use ($router) {
        $router->post('/create', [
            'uses' => 'UserController@create',
        ]);
        $router->get('/detail/{id}', [
            'uses' => 'UserController@detail',
        ]);
});

**Reply by dimaspurwadi** [reply]
ah thanks, i dont know why my IDE not import automatically.
btw can we include prefix in group? previously in my php laravel route, we can give prefix, for example:
```
$router->group(['prefix' => 'users'], function () use ($router) {
      $router->post('/create', [
            'uses' => 'UserController@create',
       ]);
       $router->get('/detail/{id}', [
             'uses' => 'UserController@detail',
       ]);
});
```

**Reply by dimaspurwadi** [reply]
hi, can anyone help me please?

**Reply by hwbrzzl** [reply]
Sorry, missed this message, like this: `facades.Route().Prefix("a").Group()`

**Reply by dimaspurwadi** [reply]
```
facades.Route().Prefix("/test").Group(func(router route.Router) {
		facades.Route().Get("/tus", func(ctx http.Context) http.Response {
			return ctx.Response().View().Make("welcome.tmpl", map[string]any{
				"version":     support.Version,
				"test_config": "hahai",
			})
		})
	})
```

please check my route, that code not work, only GET /tus available on my application.
my expectation is GET /test/tus

**Reply by hwbrzzl** [reply]
`facades.Route().Get("/tus", func(ctx http.Context) http.Response {` -> `router.Get("/tus", func(ctx http.Context) http.Response {`

**Reply by dimaspurwadi** [reply]
thanks.
i need to get {{URL::asset()}} like in laravel, how to do it in goravel?

**Reply by hwbrzzl** [reply]
The function can't be used in template directly, you need to set a variable in controller.

**Comment by tigawanna** [comment]
any openAPI and swagger middle ware?

**Reply by hwbrzzl** [reply]
Sorry, missed this message, for example?

**Comment by anammaulana** [comment]
please ajari aku goravel masih pemula saya

**Reply by hwbrzzl** [reply]
Sure, you can join the telegram group.

**Reply by anammaulana** [reply]
okey sir thks

**Comment by jeff87218** [comment]
您好，路由分組可以像Laravel那樣多層嵌套嗎?
例如
```go
facades.Route().Prefix("api").Group(func(router route.Router) {
	facades.Route().Prefix("nesting").Group(func(router route.Router) {
		router.Get("/test", testController.Index)
	})
})
```

**Reply by hwbrzzl** [reply]
支持的，但在子嵌套中需要用 `router.Prefix()`，不可使用 `facades.Route().Prefix()`。

**Reply by jeff87218** [reply]
好的 非常感謝您!

**Comment by klphp** [comment]
不支持Json格式请求吗

**Reply by hwbrzzl** [reply]
嗯？Json 请求是最基础的啊，不知道你具体遇到了什么问题？

**Reply by klphp** [reply]
获取不到提交的数据，报错如下：
```
[2024-08-08 03:58:33] local.error: <nil>
trace:
        log.(*Writer).withStackTrace:/Volumes/server/go/pkg/mod/github.com/goravel/framework@v1.14.3/log/logrus_writer.go:200
        log.(*Writer).Error:/Volumes/server/go/pkg/mod/github.com/goravel/framework@v1.14.3/log/logrus_writer.go:98
        gin.NewContextRequest:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_request.go:37
        gin.(*Context).Request:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context.go:29
        controllers.(*ValidationController).Json:/Volumes/server/go/src/goravel_example/app/http/controllers/validation_controller.go:49
        gin.(*Group).Post.handlerToGinHandler.func1:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/utils.go:30
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.(*ContextRequest).Next:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_request.go:173
        gin.(*Route).GlobalMiddleware.ResponseMiddleware.func3:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_response.go:131
        gin.(*Group).getRoutesWithMiddlewares.middlewaresToGinHandlers.middlewareToGinHandler.func3:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/utils.go:38
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.(*ContextRequest).Next:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_request.go:173
        gin.(*Route).GlobalMiddleware.Tls.func2:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/tls.go:18
        gin.(*Route).GlobalMiddleware.middlewaresToGinHandlers.middlewareToGinHandler.func4:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/utils.go:38
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.(*ContextRequest).Next:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_request.go:173
        gin.(*Route).GlobalMiddleware.Cors.func1:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/cors.go:63
        gin.(*Route).GlobalMiddleware.middlewaresToGinHandlers.middlewareToGinHandler.func4:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/utils.go:38
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.(*ContextRequest).Next:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/context_request.go:173
        providers.(*RouteServiceProvider).Boot.Kernel.Middleware.StartSession.func1:/Volumes/server/go/pkg/mod/github.com/goravel/framework@v1.14.3/session/middleware/start_session.go:60
        gin.(*Route).GlobalMiddleware.middlewaresToGinHandlers.middlewareToGinHandler.func4:/Volumes/server/go/pkg/mod/github.com/goravel/gin@v1.2.2/utils.go:38
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.LoggerWithConfig.func1:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/logger.go:249
        gin.(*Context).Next:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/context.go:185
        gin.(*Engine).handleHTTPRequest:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/gin.go:633
        gin.(*Engine).ServeHTTP:/Volumes/server/go/pkg/mod/github.com/gin-gonic/gin@v1.10.0/gin.go:589
        gin.(*Route).Run.AllowQuerySemicolons.func1:/Volumes/server/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.darwin-amd64/src/net/http/server.go:3159
        http.HandlerFunc.ServeHTTP:/Volumes/server/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.darwin-amd64/src/net/http/server.go:2166
        http.serverHandler.ServeHTTP:/Volumes/server/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.darwin-amd64/src/net/http/server.go:3137

```

**Reply by xx1196** [reply]
do u can publish your controller?

**Reply by klphp** [reply]
```
//request 
package requests

import (
	"github.com/goravel/framework/contracts/http"
	"github.com/goravel/framework/contracts/validation"
)

type UserCreateRequest struct {
	Username string `json:"username" form:"username"`
	Password string `json:"password" form:"password"`
}


func (r *UserCreateRequest) Authorize(ctx http.Context) error {
	return nil
}

func (r *UserCreateRequest) Rules(ctx http.Context) map[string]string {
	return map[string]string{
		"username": "required|max_len:255|username_exists",
		"password": "required",
	}
}

func (r *UserCreateRequest) Messages(ctx http.Context) map[string]string {
	return map[string]string{
		"username.required": "帐号不能为空",
		"password.required": "密码不能为空",
	}
}

func (r *UserCreateRequest) Attributes(ctx http.Context) map[string]string {
	return map[string]string{}
}

func (r *UserCreateRequest) PrepareForValidation(ctx http.Context, data validation.Data) error {
	return nil
}

//controller
func (r *UserController) Create(ctx http.Context) http.Response {

	var formData requests.UserCreateRequest

	return ctx.Response().Success().Json(http.Json{
		"username": formData.Username,
		"password": formData.Password,
	})
}

//curl
curl --request POST \
>   --url http://127.0.0.1:3003/users/create \
>   --header 'Content-Type: application/json' \
>   --header 'content-type: application/json' \
>   --data '{
>     "username": "abc",
>     "password":"123321"
> }'

//return
{"password":"","username":""}

```
现在没有报错了，但是返回是空的。

**Reply by klphp** [reply]
啊，	_, _ = ctx.Request().ValidateRequest(&formData)加上这句可以了。

**Reply by hwbrzzl** [reply]
Yes, please check the document: https://www.goravel.dev/zh/the-basics/validation.html#%E9%AA%8C%E8%AF%81%E8%A1%A8%E5%8D%95%E8%AF%B7%E6%B1%82

**Comment by xx1196** [comment]
hi, i have this cors https://www.goravel.dev/the-basics/routing.html#cross-origin-resource-sharing-cors, but my frontend hace the cors error, why?

Access to fetch at 'http://localhost:3001/id-types' from origin 'http://mi-dominio-local.test' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Reply by hwbrzzl** [reply]
Does your CORS as shown below?

<img width="785" alt="image" src="https://github.com/user-attachments/assets/2e978c74-00b2-4097-8ff3-826c1be174eb">

**Reply by xx1196** [reply]
@hwbrzzl yes is the same, but I get this error, I can't test my development environment for this point :'(

**Comment by halfstring** [comment]
facades.Route().Resource("users", controllers.NewUserController())

可以为上述Resource添加->except(['PATCH'])或者only()么？

**Reply by hwbrzzl** [reply]
en，未来会增加，但优先级不高，急用的话欢迎 PR 👍

**Comment by tekpriest** [comment]
is it possible to specify which http engine when using `facades.Route()?`

**Reply by hwbrzzl** [reply]
No, do you want to use gin and fiber simultaneously?

**Comment by tekpriest** [comment]
yes, fiber doesn't work well with swagger document at the moment and gin
doesn't support true routing wildcards

**Reply by hwbrzzl** [reply]
Okay, only support to use one driver for now.

**Reply by devhaozi** [reply]
Fiber also support swagger, use https://github.com/gofiber/swagger

```go
swagger.New(swagger.Config{
    Title: "Swagger",
})(ctx.(*fiber.Context).Instance())
```

![1c9dc27ed51a6636fbc513dccb14bedc](https://github.com/user-attachments/assets/a1a2474e-ad64-4928-86a4-736433ad0ab3)

Don't forget to load your docs first.

**Comment by GabbyMrH** [comment]
你这个路由分组，跟路由分组有几毛钱关系呢？

**Comment by tekpriest** [comment]
how can I make sure the server doesn't crash due to an error, but instead recovers

**Comment by tekpriest** [comment]
the server shutdowns everytime there is an uncaught error, how can I add a recover option or a global middleware to avoid

**Reply by hwbrzzl** [reply]
Could you share your screenshot?

**Reply by tekpriest** [reply]
![image](https://github.com/user-attachments/assets/e7e49585-f646-430c-ae6b-4115a0be3ea0)

**Reply by tekpriest** [reply]
just noticed now it restarts the server in non local environment

**Reply by hwbrzzl** [reply]
Hi, you can add a custom global middleware for now, now it will be integrated in 1.15.

```go
// app/http/middleware/recover.go
func Recover() httpcontract.Middleware {
	return func(ctx httpcontract.Context) {
		defer func() {
			if err := recover(); err != nil {
				ctx.Response().Json(500, httpcontract.Json{
					"error": err,
				}).Render()
			}
		}()

		ctx.Request().Next()
	}
}

// app/http/kernel.go
package http

import (
	"github.com/goravel/framework/contracts/http"
	httpmiddleware "github.com/goravel/framework/http/middleware"
	sessionmiddleware "github.com/goravel/framework/session/middleware"

	"goravel/app/http/middleware"
)

type Kernel struct {
}

// The application's global HTTP middleware stack.
// These middleware are run during every request to your application.
func (kernel Kernel) Middleware() []http.Middleware {
	return []http.Middleware{
		httpmiddleware.Throttle("global"),
		sessionmiddleware.StartSession(),
		middleware.Recover(),
	}
}
```

**Comment by klphp** [comment]
怎么获取所有路由列表

**Reply by hwbrzzl** [reply]
目前还没有显式的方式可以获取，可否描述下你的需求，我可以建一个 Issue。

**Reply by klphp** [reply]
想用路由规则搞个rbac权限管理

**Reply by hwbrzzl** [reply]
https://www.goravel.dev/zh/security/authorization.html 配合路由中间件无法满足需求吗？

**Comment by merouanemc** [comment]
There any way to return a go html template as response for the Rate limiting ?

```go
facades.RateLimiter().For("global", func(ctx http.Context) http.Limit {
  return limit.PerMinute(1000).Response(func(ctx http.Context) {
    ctx.Request().AbortWithStatus(http.StatusTooManyRequests)
// or 
ctx.Request().AbortWithStatusJson(gohttp.StatusTooManyRequests, contractshttp.Json{
				"message": "Too many requests",
			})
  })
})
```

**Reply by hwbrzzl** [reply]
You can do it like below:

```go
facades.RateLimiter().For("global", func(ctx contractshttp.Context) contractshttp.Limit {
  return limit.PerMinute(1).Response(func(ctx contractshttp.Context) {
    ctx.Response().View().Make("admin/index.tmpl").Render()
    ctx.Request().Abort(contractshttp.StatusOK)
  })
})

// or
facades.RateLimiter().For("global", func(ctx contractshttp.Context) contractshttp.Limit {
  return limit.PerMinute(1).Response(func(ctx contractshttp.Context) {
    ctx.Response().Redirect(contractshttp.StatusMovedPermanently, "view").Abort()
  })
})
```

**Comment by nguyen-truong-nhan** [comment]
ReadTimeout:  30 * time.Second,
            WriteTimeout: 30 * time.Second,
            IdleTimeout:  120 * time.Second, how to add this into config ?

**Reply by hwbrzzl** [reply]
You can add them directly, I think there's no special.

**Reply by nguyen-truong-nhan** [reply]
can you point me to example ? im using goravel/framework v1.14.9

**Reply by hwbrzzl** [reply]
Just like other configurations, do you encounter any problems?

**Reply by nguyen-truong-nhan** [reply]
just ididnt find example how to add it

**Reply by nguyen-truong-nhan** [reply]
config.Add("http", map[string]any{
		// HTTP Driver
		"default": "gin",
		// HTTP Drivers
		"drivers": map[string]any{
			"gin": map[string]any{
				// Optional, default is 4096 KB
				"body_limit":   4096,
				"header_limit": 4096,
				"route": func() (route.Route, error) {
					return ginfacades.Route("gin"), nil
				},
				// Optional, default is http/template
				"template": func() (render.HTMLRender, error) {
					return gin.DefaultTemplate()
				},
			},
		},
		// HTTP URL
		"url": config.Env("APP_URL", "http://localhost"),
		// HTTP Host
		"host": config.Env("APP_HOST", "127.0.0.1"),
		// HTTP Port
		"port": config.Env("APP_PORT", "3000"),
		// HTTPS Configuration
		"tls": map[string]any{
			// HTTPS Host
			"host": config.Env("APP_HOST", "127.0.0.1"),
			// HTTPS Port
			"port": config.Env("APP_PORT", "3000"),
			// SSL Certificate, you can put the certificate in /public folder
			"ssl": map[string]any{
				// ca.pem
				"cert": "",
				// ca.key
				"key": "",
			},
		},
	}) this is config but with these config not sure where to add inside gin block or where ?

**Reply by hwbrzzl** [reply]
Do you want to inject ReadTimeout, etc. when initiating the `gin` instance? If so, it's not supported yet. Welcome to submit a PR to goravel/gin.
